sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/OrderedListUserState.java

```diff
@@ -0,0 +1,240 @@
+    // accessible in pre-existing iterables even after:
+    //   (1) a sort key is added to or removed from pendingAdds, or
+    //   (2) a new value is added to an existing sort key
+    ArrayList<PrefetchableIterable<TimestampedValue<T>>> pendingAddsInRange = new ArrayList<>();
+    for (Entry<Instant, Collection<T>> kv :
+        pendingAdds.subMap(minTimestamp, limitTimestamp).entrySet()) {
+      pendingAddsInRange.add(
+          PrefetchableIterables.limit(
+              Iterables.transform(kv.getValue(), (v) -> TimestampedValue.of(v, kv.getKey())),
+              kv.getValue().size()));
+    }
+    Iterable<TimestampedValue<T>> valuesInRange = Iterables.concat(pendingAddsInRange);
+
+    if (!isCleared) {
+      StateRequest.Builder getRequestBuilder = this.request.toBuilder();
+      getRequestBuilder
+          .getStateKeyBuilder()
+          .getOrderedListUserStateBuilder()
+          .getRangeBuilder()
+          .setStart(minTimestamp.getMillis())
+          .setEnd(limitTimestamp.getMillis());
+
+      // TODO: consider use cache here
+      CachingStateIterable<TimestampedValue<T>> persistentValues =
+          StateFetchingIterators.readAllAndDecodeStartingFrom(
+              Caches.noop(),
+              this.beamFnStateClient,
+              getRequestBuilder.build(),
+              this.timestampedValueCoder);
+
+      // Make a snapshot of the current pendingRemoves and use them to filter persistent values.
+      // The values of pendingRemoves are kept, so that they will still be accessible in
+      // pre-existing iterables even after a sort key is removed.
+      TreeRangeSet<Instant> pendingRemovesSnapshot = TreeRangeSet.create(pendingRemoves);
+      Iterable<TimestampedValue<T>> persistentValuesAfterRemoval =
+          Iterables.filter(
+              persistentValues, v -> !pendingRemovesSnapshot.contains(v.getTimestamp()));
+
+      return Iterables.mergeSorted(
+          ImmutableList.of(persistentValuesAfterRemoval, valuesInRange),
+          Comparator.comparing(TimestampedValue::getTimestamp));
+    }
+
+    return valuesInRange;
+  }
+
+  public Iterable<TimestampedValue<T>> read() {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    return readRange(Instant.ofEpochMilli(Long.MIN_VALUE), Instant.ofEpochMilli(Long.MAX_VALUE));
+  }
+
+  public void clearRange(Instant minTimestamp, Instant limitTimestamp) {
+    checkState(
+        !isClosed,
+        "OrderedList user state is no longer usable because it is closed for %s",
+        request.getStateKey());
+
+    // Remove items (in a collection) in the specific range from pendingAdds.
+    // The old values of the removed sub map are kept, so that they will still be accessible in
+    // pre-existing iterables even after the sort key is cleared.
+    pendingAdds.subMap(minTimestamp, true, limitTimestamp, false).clear();
```

Comment:
I think this is the same as just calling `.subMap(minTimestamp, limitTimestamp)`.

Reply:
You are right. Will make the change.

Reply:
Done.